Forge client support (support integrating with node-minecraft-protocol-forge)

deathcap opened the pull request:
Add client-side https://github.com/PrismarineJS/node-minecraft-protocol/issues/114 Forge Support

* [x] Send tagged hostname in set_protocol packet
* [x] Receive server channel registrations (FML|HS, FML, FML|MP, FML, FORGE)
* [x] Handle the FML|HS ServerHello with the player's dimension
* [x] Register client plugin channels (FML|HS, FML, FML|MP, FML, FORGE)
* [x] Send the FML|HS ClientHello
* [x] Send client modList
* [x] Receive server modList
* [x] Check client/server mod compatibility - emits `forgeMods` event for app to check
* [x] Send HandshakeAck
* [x] Receive RegistryData packet(s) until hasMore=false
* [x] Send/receive final HandshakeAcks and CompleteHandshake
* [x] Refactor discriminator byte / 'FMLIndexedMessageToMessageCodec' to use a mapper
* [x] Refactor discriminator byte / 'FMLIndexedMessageToMessageCodec' parsing to use one big switch
* [x] Move FML|HS handshake to src/fml.js, support and document forge/forgeMods opts, event
* [x] Refactor into separate module?
* [x] Update for https://github.com/PrismarineJS/node-minecraft-protocol/pull/333 Modularization attempt of createClient()
* [x] Move tagHost=\0FML\0 to be installed by src/client/forgeHandshake.js, not the client example

Not included, out of scope for this PR:
* Fully parse the RegistryData packets and emit registry events - would be nice but only needed for advanced usage, can come later
* FORGE plugin channel (observed in packet logging, sends lava and water)
* FML and FML|MP plugin channels (registered, but not seen that they are used at least at login)
* Server-side Forge support https://github.com/PrismarineJS/node-minecraft-protocol/issues/114

deathcap commented:
Current status (testing against server: forge-1.8.9-11.15.0.1715 + mods/ironchest-1.8.9-6.0.121.768.jar):

```
node-minecraft-protocol $ NODE_DEBUG=mc-proto node examples/client_forge/client_forge.js localhost 25565
MC-PROTO: 19450 writing packet handshaking.set_protocol
MC-PROTO: 19450 { protocolVersion: 47,
  serverHost: 'localhost\u0000FML\u0000',
  serverPort: 25565,
  nextState: 2 }
MC-PROTO: 19450 writing packet login.login_start
MC-PROTO: 19450 { username: 'echo' }
connected
MC-PROTO: 19450 read packet login.compress
MC-PROTO: 19450 { threshold: 256 }
MC-PROTO: 19450 read packet login.success
MC-PROTO: 19450 { uuid: 'c39fa4c6-85a3-3582-8167-db3805dbe851',
  username: 'echo' }
MC-PROTO: 19450 read packet play.custom_payload
MC-PROTO: 19450 { channel: 'REGISTER',
  data: <Buffer 46 4d 4c 7c 48 53 00 46 4d 4c 00 46 4d 4c 7c 4d 50 00 46 4d 4c 00 46 4f 52 47 45> }
MC-PROTO: 19450 read packet play.custom_payload
MC-PROTO: 19450 { channel: 'FML|HS', data: <Buffer 00 02 00 00 00 00> }
```

No longer kicked, but need to handle the plugin channels and FML handshakes

deathcap commented:
Getting further, to the mod list exchange:

```
client_forge $ NODE_DEBUG=mc-proto node client_forge.js localhost 25565
MC-PROTO: 24794 writing packet handshaking.set_protocol
MC-PROTO: 24794 { protocolVersion: 47,
  serverHost: 'localhost\u0000FML\u0000',
  serverPort: 25565,
  nextState: 2 }
MC-PROTO: 24794 writing packet login.login_start
MC-PROTO: 24794 { username: 'echo' }
connected
MC-PROTO: 24794 read packet login.compress
MC-PROTO: 24794 { threshold: 256 }
MC-PROTO: 24794 read packet login.success
MC-PROTO: 24794 { uuid: 'c39fa4c6-85a3-3582-8167-db3805dbe851',
  username: 'echo' }
MC-PROTO: 24794 read packet play.custom_payload
MC-PROTO: 24794 { channel: 'REGISTER',
  data: <Buffer 46 4d 4c 7c 48 53 00 46 4d 4c 00 46 4d 4c 7c 4d 50 00 46 4f 52 47 45 00 46 4d 4c> }
Server-side registered channels: [ 'FML|HS', 'FML', 'FML|MP', 'FORGE', 'FML' ]
MC-PROTO: 24794 read packet play.custom_payload
MC-PROTO: 24794 { channel: 'FML|HS', data: <Buffer 00 02 00 00 00 00> }
FML|HS { data: 
   { discriminator: 0,
     fmlProtocolVersionServer: 2,
     overrideDimension: 0,
     fmlProtocolVersionClient: undefined,
     mods: undefined },
  metadata: { size: 6 },
  buffer: <Buffer 00 02 00 00 00 00> }
MC-PROTO: 24794 writing packet play.custom_payload
MC-PROTO: 24794 { channel: 'REGISTER',
  data: <Buffer 46 4d 4c 7c 48 53 00 46 4d 4c 00 46 4d 4c 7c 4d 50 00 46 4d 4c 00 46 4f 52 47 45> }
MC-PROTO: 24794 writing packet play.custom_payload
MC-PROTO: 24794 { channel: 'FML|HS', data: <Buffer 01 02> }
MC-PROTO: 24794 writing packet play.custom_payload
MC-PROTO: 24794 { channel: 'FML|HS', data: <Buffer 02 00> }
MC-PROTO: 24794 read packet play.kick_disconnect
MC-PROTO: 24794 { reason: '"Mod rejections [FMLMod:IronChest{6.0.121.768}]"' }
```

The mod list will likely need to come from the ServerListPing, since a client must match the server and does not really need to implement all of the mods (for most bots/scripts/clients expected as the use case of NMP). But if I send:

```javascript
         mods: [
           {name:'IronChest', version:'6.0.121.768'}
         ]
```

get an error encoding the mods array:

```
/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/protodef.js:157
        throw e;
        ^

Error: SizeOf error for mods.2.0.name : missing data type: string
    at ProtoDef.sizeOf (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/protodef.js:140:15)
    at /Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:153:21
    at tryCatch (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/utils.js:33:12)
    at tryDoc (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/utils.js:40:10)
    at /Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:152:19
    at Array.reduce (native)
    at ProtoDef.sizeOfContainer (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:148:23)
    at ProtoDef.sizeOf (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/protodef.js:143:33)
    at /Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/datatypes/structures.js:96:28
    at tryCatch (/Users/admin/games/voxeljs/node-minecraft-protocol/node_modules/protodef/dist/utils.js:33:12)
```

deathcap commented:
```javascript
Server ModList: [ { name: 'Forge', version: '11.15.0.1715' },
  { name: 'IronChest', version: '6.0.121.768' },
  { name: 'mcp', version: '9.18' },
  { name: 'FML', version: '8.0.99.99' } ]
```

rom1504 commented on examples/client_forge/client_forge.js:
```diff
@@ -0,0 +1,191 @@
+        "type": "byte"
+      },
+
+      // ServerHello
```
(as I was saying in gitter), it might make more sense to use a big switch+ a mapper, like in https://github.com/PrismarineJS/prismarine-nbt/blob/master/nbt.json#L48 (or https://github.com/PrismarineJS/node-minecraft-protocol/blob/master/src/transforms/serializer.js#L21 )
* mapper makes it possible for user to use the names (for example ServerHello instead of 0)
* the big switch probably makes more sense because fields are distinct between each kind of "packet" here

deathcap replied:
agreed, that's what I wanted to do but couldn't figure out how, to do: investigate those references

deathcap commented:
Server debug logging for reference: https://gist.github.com/deathcap/2c31e498b2375053c0e0

The handshake completes, but exits immediately after; still need to hook up the "handshake complete state" to node-minecraft-protocol's notion of handshake/login complete, and also general refactoring

deathcap commented:
This is now functional, just needs to be cleaned up/refactored

deathcap commented:
Removing [WIP], further enhancements are possible but I think this is working well enough to review, let me know what you think. Future changes I would like but not included here:

* Move `FML|HS` protodef to [minecraft-data](https://github.com/PrismarineJS/minecraft-data) - and other plugin channel payloads like [MC|AdvCdm] (https://github.com/PrismarineJS/mineflayer/commit/24f24349261c5b5031ff4ec624fef670af3028a5) and [MC|Brand](https://github.com/PrismarineJS/mineflayer/pull/254) in mineflayer, requires 
* Add a generic plugin channel API to node-minecraft-protocol (and mineflayer: https://github.com/PrismarineJS/mineflayer/issues/361)), right now I just listen for `custom_payload` events
* Move the pre-connect ping into createClient() / SRV refactoring - https://github.com/PrismarineJS/node-minecraft-protocol/issues/327 - right now it is in the example, but would be more convenient for apps if it was built-in to the library (automatically support bots connecting to Forge servers with no changes on their part), except it is a disruptive change so I left it out for now
* More parsing of `RegistryData` and more events - a fully-functional FML/Forge client would need these, but for most bots and simple clients, it is not needed, can come later - this PR gets handshaking/connecting working, a necessary prerequisite for even connecting to modded servers

rom1504 commented:
Roblabla was talking about putting this as a plugin in an other repo. It would probably be possible and it raises the question of how much non-vanilla we want to support directly it nmp. It might be better to put these kind of extensions in their own module so it can be independently maintained. Especially if it's needed to support lot of forge mods.
Same idea for Minecraft-data : do we want to put non vanilla content in there or should it be in its own repo ? Not sure

Putting everything in nmp and mcdata themselves has the benefit of making it a bit easier for the users, but it means a bit more code to maintain.

Apart from where the code should go, it looks fine to me.

deathcap commented:
Alright I'll see about refactoring out into a separate module. There may be some layering difficulties though, FML requires the `set_protocol` packet serverHost to end in `\0FML\0`, I can pass this to createClient() but not sure how DNS SRV resolution will react to trying to resolve it, need to test further, think about how/if NMP could be enhanced to support an optional FML|HS plugin

deathcap commented:
Started factoring out in https://github.com/deathcap/node-minecraft-protocol-forge - works with the 0.16.6 minecraft-protocol release, but some concerns:

NMP Resolving \0\FML\0 DNS Hostnames

As part of the FML protocol I append a null byte and the FML string and another null byte to the serverHost in the set_protocol packet, to get this to work with no changes to the NMP release, I just pass this 'tagged' hostname to the `host` parameter of NMP's `createClient()`. It does seem to work, at least on my system — I imagine because null is recognized as a terminator on the underlying network APIs, e.g.:

```javascript
> require('dns').resolve('github.com', function(){console.log(arguments)})
QueryReqWrap {
  bindingName: 'queryA',
  callback: { [Function: asyncCallback] immediately: true },
  hostname: 'github.com',
  oncomplete: [Function: onresolve],
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> { '0': null, '1': [ '192.30.252.131' ] }

> require('dns').resolve('github.com\0FML\0', function(){console.log(arguments)})
QueryReqWrap {
  bindingName: 'queryA',
  callback: { [Function: asyncCallback] immediately: true },
  hostname: 'github.com\u0000FML\u0000',
  oncomplete: [Function: onresolve],
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> { '0': null, '1': [ '192.30.252.128' ] }
```

but its not clear if it is a good idea to rely on this behavior. Another option would be for NMP to support another option to append a string to the hostname, but only in the set_protocol packet, see: https://github.com/deathcap/node-minecraft-protocol/commit/93c5c7315a81a379906c0df9862d04d9b9cd5c0a

Second is the larger issue of how to cleanly integrate NMP with NMPF. The client has to advertise the fact that it supports FML/Forge, so I [wrap the createClient()](https://github.com/deathcap/node-minecraft-protocol-forge/commit/16fa0023a3eeddcdcc069fa3a3829aa8f53f17ab) function from NMP, to add the tag (or alter the hostname, for now).

deathcap commented:
Also added [node-minecraft-protocol-forge](https://github.com/deathcap/node-minecraft-protocol-forge) support to [node-minecraft-protocol-auto](https://github.com/deathcap/node-minecraft-protocol-auto), but to be honest I'd still favor integrating either or both into node-minecraft-protocol proper.

What about a compromise, add -forge and -auto as separate "modules" but in this same repository? Along the lines of https://github.com/PrismarineJS/node-minecraft-protocol/pull/210 [WIP] Modularize NMP. As long as they are cleanly separated from the rest of the code, I think it should not increase maintenance cost too much, or if they do become a problem, could disable the modules without affecting the rest of the project.

rom1504 commented:
Hmm, I think it should be used as a "plugin" of node-minecraft-protocol, and not as an independant createClient. See also:
> 13:49 < roblabla> @deathcap those using forge know they are using forge
> 13:51 < roblabla> What I'd really like is finishing the NMP modularisation work, so we could just do
 ```
createClient({
modules: [ require('nmp-forge'), "handshake", "encryption", "compression", "play" ]
 });
```
> 13:51 < roblabla> Which would inject the code required to connect on a forge server

deathcap commented:
I guess I can wait until GH-210 Modularize NMP to update this code to be a plugin of node-minecraft-protocol. Features needed:

* ability to append to the hostname in the `set_protocol` packet
* ability to use plugin channels (possible already by listening on `custom_payload` packets, but ofc GH-328 plugin channel API would be cleaner)
* ability to accept new client options and emit new client events

deathcap commented:
Updated for GH-333 client modularization, but an outstanding issue is having the client_forge example need to set tagHost \0FML\0. This logic should not need to be known by the app, ought to be refactored into the src/client/forgeHandshake.js middleware.

But if I have src/client/forgeHandshake.js set options.tagHost, then it is too late for src/client/setProtocol.js to read it, so the server doesn't know the client supports Forge. Likely the same problem (and solution) described in https://github.com/PrismarineJS/node-minecraft-protocol/pull/331#issuecomment-174941928 - may need some kind of 'pause' mechanism, to change the client object before it is fully created and ready (`forgeHandshake(client, options)` is called immediately after `createClient()`, but it is still too late).

deathcap commented:
Moved tagHost \0FML\0 into the plugin. It is now enabled when `forgeMods` is set, in src/createClient.js, because it has to be set before setProtocol (since it is the plugin that sends the serverHost). With these changes, I believe this PR is ready.

Note: the `tagHost` option added to src/client/setProtocol.js could also be used for https://github.com/PrismarineJS/node-minecraft-protocol/issues/336 BungeeCord protocol support

Note 2: the client_forge example demonstrates auto-negotiation, since that's the likely usage for Forge clients written with node-minecraft-protocol, but it could be refactored after https://github.com/PrismarineJS/node-minecraft-protocol/pull/331 dynamic protocol support. Other enhancements/cleanup are possible, but I think this PR is good enough to pull at this point.

rom1504 commented:
Looks good to me, but I think it belongs in its own repo.
> 09:39 < rom1504> deathcap: https://github.com/PrismarineJS/node-minecraft-protocol/pull/326/files could be organized as https://github.com/andrewrk/mineflayer-navigate/ : forgeHandshake.js could be the index.js and client_forge.js could be the example.js
> 09:39 < rom1504> ah and `if (options.forgeMods) forgeHandshake(client, options);` should be in the client_forge.js / example.js imho

deathcap commented:
Updated to fix conflicts with master but still needs to be moved into its own repo, re:gitter, agreed could be best served in a repository under the PrismarineJS organization in order to ease maintenance. src/client/forgeHandshake.js is easy to separate but will need to untangle the tagHost and autoVersion.

rom1504 commented:
@deathcap you should now be able to create a repo in PrismarineJS for that forge plugin . (Or move https://github.com/deathcap/node-minecraft-protocol-forge but then I think you should first remove almost all the commits, it's a plugin, not a nmp fork)

deathcap commented:
Cool thanks. Yeah https://github.com/deathcap/node-minecraft-protocol-forge branched from this repo (based on this PR) but I should probably rebase to remove the pre-Forge commits which were since removed as part of moving to a new module, or save the old branch and make a new one just with src/client/forgeHandshake.js

deathcap commented:
src/client/autoVersion.js has a dependency on src/client/forgeHandshake.js, though. I suppose it could be modified to support "hooks" into the response modinfo, which forgeHandshake could register itself as a handler somehow (for response.modinfo.type === 'FML', install itself with forgeMods: response.info.modList).

rom1504 commented:
Yes it would be better to remove that dependency somehow

deathcap commented:
Well it's easy enough to remove the autoVersion/forgeHandshake dependency, but it would remove the feature :). The trickiness is in having the plugin add Forge support to autoVersion, without necessarily enabling it in the handshake. Maybe the node-Minecraft-protocol-forge module could expose two plugins (forgeHandshake and autoVersionForge?)

deathcap commented:
Alright I've rebased https://github.com/deathcap/node-minecraft-protocol-forge, and updated this PR so it now only contains the changes needed to support node-minecraft-protocol-forge rather than the code itself. To allow auto-versioning Forge plugins, added a new "auto version hook" feature, where plugins can add their own handlers for the auto-version ping response. autoVersionForge uses this to install itself with the right mods. Example usages of these two plugins:

```javascript
var mc = require('minecraft-protocol');
var forgeHandshake = require('minecraft-protocol-forge').forgeHandshake;
var client = mc.createClient({
    host: host,
    port: port,
    username: username,
    password: password
});

forgeHandshake(client, {forgeMods: [
  { modid: 'mcp', version: '9.18' },
  { modid: 'FML', version: '8.0.99.99' },
  { modid: 'Forge', version: '11.15.0.1715' },
  { modid: 'IronChest', version: '6.0.121.768' }
]);
```

and:

```javascript
var mc = require('minecraft-protocol');
var autoVersionForge = require('minecraft-protocol-forge').autoVersionForge;
var client = mc.createClient({
    version: false,
    host: host,
    port: port,
    username: username,
    password: password
});

autoVersionForge(client);
```

Not bad, imho. Open to other improvements but once these changes are in, will be able to publish node-minecraft-protocol-forge.

@rom1504 
> @deathcap you should now be able to create a repo in PrismarineJS for that forge plugin 

OK I tried to create a new repo in PrismarineJS but am only offered:

<img width="429" alt="screen shot 2016-02-07 at 12 01 26 am" src="https://cloud.githubusercontent.com/assets/5897956/12871415/faf8732a-cd2d-11e5-84d4-c69685f1c7af.png">

can I just transfer https://github.com/deathcap/node-minecraft-protocol-forge to https://github.com/PrismarineJS/node-minecraft-protocol-forge? (there is a 'transfer' option in https://github.com/deathcap/node-minecraft-protocol-forge/settings but it is in the 'danger zone' so wanted to be sure)

rom1504 commented:
Yeah just use transfer.

rom1504 commented:
Looks good to me. Is this ready to merge ?

rom1504 commented:
Oh and you probably got an invitation to join prismarinejs in your mail, looks like you didn't accept it yet.

deathcap commented:
Yep ready to merge!

I'll check for the invite, won't let me transfer to PrismarineJS without it (says no rights)
